src/java.base/share/classes/java/lang/Thread.java

```diff
@@ -1805,14 +1805,17 @@ boolean getAndClearInterrupt() {
         if (com.ibm.oti.vm.VM.isJVMInSingleThreadedMode()) {
             return interruptedImpl();
         }
-        synchronized (interruptLock) {
-            boolean oldValue = interrupted;
-            if (oldValue) {
-                interrupted = false;
-                clearInterruptEvent();
+        boolean oldValue = interrupted;
+        if (oldValue) {
+            synchronized (interruptLock) {
+                oldValue = interrupted;
+                if (oldValue) {
+                    interrupted = false;
+                    clearInterruptEvent();
+                }
             }
```

Comment:
In the equivalent `VirtualThread.java` method, the `synchronized` block is just within the if-statement:
https://github.com/ibmruntimes/openj9-openjdk-jdk21/blob/cd40810ef43b17fdd5c554b0f5f4772a1ccd1a0a/src/java.base/share/classes/java/lang/VirtualThread.java#L885-L895

Is there a reason as to why that approach wouldn't work here?

Reply:
Functionally, no difference will be seen. In the case, where there is high contention i.e. multiple threads attempt to enter the `synchronized` block and clear the interrupt, only one thread will invoke the native method (`Thread.interruptedImpl`) due to the extra `if` statement. Otherwise, all threads that enter the `synchronized` block will invoke the native method; and this will lower the throughput when resources are scarce. At the cost of an extra `if` statement, we lower the invocations of the native method in the worst case to improve perf.

Reply:
If there is such a performance benefit, we should likely apply this to the `VirtualThread.java` method too.

Reply:
Waiting for perf data. Will apply the change to `VirtualThread.java` after reviewing perf data.

Reply:
- Related lock contention is completely removed. There is a 15% perf improvement.
- Also, VirtualThread.java has been updated.